### ISSUE TITLE
balena-cli phones home all commands and PII

### ISSUE THREAD

**Issue by kazazes** [balena-io/balena-cli#1359]
- **balena CLI version:** 11.7.2
- **Operating system version:** macOS
- **32/64 bit OS and processor:** 64-bit
- **Install method:** Exec

---

I noticed a small delay after running any command, starting with `balena help`. (Point 1: your tracking is slowing down your cli tool...)

Upon closer review, @resinio is sending [**every command and user** info?!](https://github.com/balena-io/balena-cli/blob/062fadfa498802a454407a8fdf53d7d8e7b7961d/lib/events.ts#L53-L61) straight to their hosted mixpanel.

It happens on [every single command.](https://github.com/balena-io/balena-cli/blob/7e1f4791ed1bb6afc4fc642b0c72890e457c5a8e/lib/app-capitano.coffee#L157-L158) Not failed commands, not hung commands, every single one.

**You're including cleartext passwords in the tracking.** Is your mixpanel setup as secure as your user authentication db? Are these passwords now accessible to anyone with access to mixpanel? 

From your docs:
`balena login --credentials --email johndoe@gmail.com --password secret`

I'm upset because I love your platform and trust you with plenty of sensitive data. This is such a poorly thought out decision.

```
return mixpanel.track(`[CLI] ${command.signature.toString()}`, {
	distinct_id: username,
	argv: process.argv.join(' '),
	version: packageJSON.version,
	node: process.version,
	arch: process.arch,
	balenaUrl,
	platform: process.platform,
	command: capitanoCli,
});
```

We trust you to run VPNs with your own certs, provision our devices, and control our SSH keys. Even Google and AWS CLI tools don't do this, in fact, I've never heard of a CLI tool doing this without disclosure, and definitely not via a third party with the full content of the command. Doesn't openBalena rely on balena-cli as well?

> With openBalena, we want to give developers the ability to manage fleets of devices on their own. One of our primary motivations in releasing openBalena is to mitigate fears of lock-in and remove barriers to exit (and entry!).

**Comment by petrosagg** [comment]
> I'm upset because I love your platform and trust you with plenty of sensitive data. This is such a poorly thought out decision.

@kazazes thanks for bringing this to our attention. First and foremost, you're absolutely right that this is an oversight and have every reason to be upset. It was never out intention to record secret values passed as CLI arguments --especially the user password-- in our analytics system. 

**Immediate actions:**
* We have fixed and released a new version of our CLI today (v11.7.3) that only records the raw command template (before variable replacement)
* We are backporting the patch to the most popular major versions
* We are analysing all our mixpanel events for occurrences of leaked secret data and we'll inform all affected users by email

> Doesn't openBalena rely on balena-cli as well?

It does, but the mixpanel configuration is dynamic and it is set to a [purposefully invalid value](https://github.com/balena-io/open-balena/blob/master/compose/services.yml#L39).  However, after looking closer at the mixpanel library implementation we realised that requests are being sent anyway to mixpanel with the invalid token.

Mixpanel most likely ignores these events but as far as security is concerned it doesn't matter. We are fixing this issue as well, but since the sent events are inaccessible we're unable to know the impact exactly. Given the beta state of openBalena though we think the impact will be relatively low.

**Why we're not removing telemetry completely**

As I mentioned above, our CLI will still report usage statistics about which commands are being used. This is valuable information for us that helps us shape decisions about features and the design of the CLI that ultimate makes a better tool for our users.

For example, we recently used usage data to conclude that a particular feature of the CLI was so little used that we're better off removing it and go back to the drawing board. Had we not have the statistics it would be hard to estimate the impact on users.

Lastly, as a response to this incident we have started discussing internally about producing a clear policy towards analytics. This is to help everyone understand what our approach to analytics is, what we track, and why. We hope that this will lead to better decisions in the future.
